src/sim/circuit.cc

```diff
@@ -0,0 +1,554 @@
+  // fprintf(fout, " %d", result);
+
+  // 2. DP, fuse gates and add shuffle gates
+  int idx = 0;
+  auto schedules = get_schedules(*seq, local_qubits, ctx);
+  for (auto &schedule : schedules) {
+    // add shuffle gate
+    std::vector<int> target;
+    std::vector<int> global;
+    for(int i = 0; i < num_qubits; i++) {
+      if (local_qubits[idx][i]){
+        target.push_back(i);
+      }
+      else {
+        global.push_back(i);
+      }
+    }
+
+    for(int i = 0; i < n_global; i++) {
+      target.push_back(global[i]);
+    }
+    Gate<DT> gate{SHUFFLE, num_qubits, 0, target, {}, {}, {}};
+    gates.push_back(gate);
+
+    schedule.compute_kernel_schedule(
+        {0, 10.4, 10.400001, 10.400002, 11, 40, 46, 66});
+    std::cout << "cost = " << schedule.cost_ << std::endl;
+    // schedule.print_kernel_schedule();
+    int num_kernels = schedule.kernels.size();
+    for (int i = 0; i < num_kernels; i++) {
+      SimGateType g_type = FUSED;
+      unsigned n_target = schedule.kernel_qubits[i].size();
+      unsigned n_control = 0;
+      std::vector<int> target;
+      std::vector<int> control;
+      // TODO: use template for matrix type, to be compatible with Legion-based version
+      // std::vector<std::complex<DT>> mat;
+      std::cout << "Fusing Kernel " << i << ": qubits [";
+      for (int j = 0; j < (int)schedule.kernel_qubits[i].size(); j++) {
+        std::cout << schedule.kernel_qubits[i][j];
+        target.push_back(schedule.kernel_qubits[i][j]);
+        if (j != (int)schedule.kernel_qubits[i].size() - 1) {
+          std::cout << ", ";
+        }
+      }
+      std::cout << "], gates ";
+      std::cout << schedule.kernels[i].to_string() << std::endl;
+      // mat = FuseGates<DT>(kernels);
+      auto mat = FuseGates(schedule.kernels[i], schedule.kernel_qubits[i], ctx);
+
+      // add fused gates kernels 
+      Gate<DT> gate{g_type,
+                    n_target,
+                    n_control,
+                    target,
+                    control,
+                    {},
+                    mat};
+      gates.push_back(gate);
+    }
+
+    idx++;
+  }
+
+  return true;
+}
+
+template <typename DT>
+void qcircuit::Circuit<DT>::simulate(int ndevices) {
+  using Simulator = SimulatorCuQuantum<DT>;
+  Simulator simulator(n_local, n_global, ndevices);
+  std::vector<unsigned> init_perm;
+  for (int i = 0; i < n_local + n_global; i++) {
+    init_perm.push_back(i);
+  }
+  simulator.InitState(init_perm);
+  printf("Init State Vectors!\n");
+  int index = 0;
+  while (index < gates.size()) {
+    Gate<DT> gate = gates[index++];
+    if (gate.gtype == SHUFFLE) {
+      simulator.ApplyShuffle(gate);
+    } else {
+      for (int i = 0; i < ndevices; i++)
+        simulator.ApplyGate(gate, i);
+    }
+  }
+  printf("Finish Simulating!\n");
+  simulator.Destroy();
+  printf("Destroyed the simulator\n");
+}
+
+/* private functiona for qcircuit::Circuit */
+// TODO: add more gates
```

Comment:
I believe there should be an automated way to do this... maybe using the `PER_GATE` macro?

Reply:
This function is not used now. It's just for some testing before integrated with quartz. We are using the quartz's gates now.